### PR TITLE
Use 16-bit random value in validator filter

### PR DESCRIPTION
**Pull request by jtraglia**
See: https://docs.google.com/spreadsheets/d/18VeMVLO1jbP8xHkT4faGEb5SyrWQhQLcU-BtsL9pTss

This fixes an oversight in `compute_proposer_index` and `get_next_sync_committee_indices`.

Currently, we use an 8-bit random byte to perform an effective balance filter. This was used to penalize validators with effective balances less than 32 ETH. But now that we have raised the maximum effective balance, this mechanism plays a more important role in validator selection.

One might expect a validator with a greater effective balance to have a greater chance of being selected as proposer or sync committee participant, but no. This is because the random byte can only represent 256 "buckets" of effective balance. So for example, currently, a validator with 33 ETH has the same likelihood of being selected as a validator with 40 ETH, because these validators are in the same bucket.

The solution is to use a bigger random value to allow more precision. I propose we use a 16-bit random value instead which provides enough precision (~65k buckets). Originally I wanted to use 32-bit values but it would require more complex changes because there would be an integer overflow when multiplying the max values together.

### CODE REVIEW

**Comment by mkalinin** [comment]
The change looks good and makes sense to me! 👍
I think we need more tests to ensure implementation compliance

**Comment by rolfyone** [comment]
Sounds logical... good writeup

**Comment by jtraglia** [comment]
> I think we need more tests to ensure implementation compliance

It shouldn't be necessary to add extra tests. There should be several sanity tests which will check this.

For example, `test_proposer_after_inactive_index` and `test_high_proposer_index` would most likely error if `get_beacon_proposer_index` were implemented incorrectly.

* https://github.com/ethereum/consensus-specs/blob/554189d105a8f4e0030a1293d73b370adb581e46/tests/core/pyspec/eth2spec/test/phase0/sanity/test_blocks.py#L669-L716

**Review by mkalinin — APPROVED** [review]
The PR looks good to me and great work on the analysis! 👍

File: specs/electra/beacon-chain.md [hunk 1]
```diff
@@ -418,23 +418,26 @@ class BeaconState(Container):
...
-    MAX_RANDOM_BYTE = 2**8 - 1
+    MAX_RANDOM_VALUE = 2**16 - 1  # [Modified in Electra]
     i = uint64(0)
     total = uint64(len(indices))
     while True:
         candidate_index = indices[compute_shuffled_index(i % total, total, seed)]
-        random_byte = hash(seed + uint_to_bytes(uint64(i // 32)))[i % 32]
+        # [Modified in Electra]
+        random_bytes = hash(seed + uint_to_bytes(i // 16))
+        offset = i % 16 * 2
```
Comment (ralexstokes) [inline specs/electra/beacon-chain.md:436]: noting we get the intended change here as precedence in python gives us this `(i % 16) * 2`

**Comment by jtraglia** [comment]
> we should verify there is some difference in test output

Good call. I have a simple difference. In `test_high_proposer_index`, the state's slot will be different. This is because the following check will require more iterations.

https://github.com/ethereum/consensus-specs/blob/98cf01d22dbb1ce2deada5dd11a5d029e54cf6a4/tests/core/pyspec/eth2spec/test/phase0/sanity/test_blocks.py#L706-L708

See `slot: 17` in the first test run (on dev) and `slot: 20` on the second one (on this branch):

image

**Pull request merged**